freqtrade/strategy/hyper.py

```diff
@@ -0,0 +1,93 @@
+"""
+IHyperStrategy interface, hyperoptable Parameter class.
+This module defines a base class for auto-hyperoptable strategies.
+"""
+from abc import ABC
+from typing import Union, List, Iterator, Tuple
+
+from skopt.space import Integer, Real, Categorical
+
+from freqtrade.strategy.interface import IStrategy
+
+
+class Parameter(object):
+    """
+    Defines a parameter that can be optimized by hyperopt.
+    """
+    default: Union[int, float, str, bool]
+    space: List[Union[int, float, str, bool]]
+    category: str
+
+    def __init__(self, *, space: List[Union[int, float, str, bool]], default: Union[int, float, str, bool] = None,
+                 category: str = None, **kwargs):
+        """
+        Initialize hyperopt-optimizable parameter.
+        :param space: Optimization space. [min, max] for ints and floats or a list of strings for categorial parameters.
+        :param default: A default value. Required for ints and floats, optional for categorial parameters (first item
+         from the space will be used). Type of default value determines skopt space used for optimization.
+        :param category: A parameter category. Can be 'buy' or 'sell'. This parameter is optional if parameter field
+         name is prefixed with 'buy_' or 'sell_'.
+        :param kwargs: Extra parameters to skopt.space.(Integer|Real|Categorical).
+        """
+        assert 'name' not in kwargs, 'Name is determined by parameter field name and can not be specified manually.'
+        self.value = default
+        self.space = space
+        self.category = category
+        self._space_params = kwargs
+        if default is None:
+            assert len(space) > 0
+            self.value = space[0]
+
+    def get_space(self, name: str) -> Union[Integer, Real, Categorical, None]:
+        """
+        Create skopt optimization space.
+        :param name: A name of parameter field.
+        :return: skopt space of this parameter, or None if parameter is not optimizable (i.e. space is set to None)
+        """
+        if not self.space:
+            return None
+        if isinstance(self.value, int):
+            assert len(self.space) == 2
+            return Integer(*self.space, name=name, **self._space_params)
+        if isinstance(self.value, float):
+            assert len(self.space) == 2
+            return Real(*self.space, name=name, **self._space_params)
+
+        assert len(self.space) > 0
+        return Categorical(self.space, name=name, **self._space_params)
+
+
+class IHyperStrategy(IStrategy, ABC):
```

Comment:
`ABC` is indeed not necessary. IDE showed a warning and this was one of the suggestions. I suppose it would be better to just add a comment that disables warning instead. Hmm mixin idea is great! I will try that 👍🏻 If it was to be a mixin class should probably be renamed too. `HyperOptable` maybe? Idk naming is hard.

Reply:
Oh i did not mention `.value` bit. Yeah seems like i missed `__int__` being a thing. That + `__float__` + `__str__` + `__bool__` seems to be all we need. Thing is i am not sure if it would work considering other side of comparison would be pandas dataframe. This needs investigation 👀

Reply:
i think it could work - HOWEVER, i also think it's too dangerous, and using `.value` is better than forcing users to use `int()`.

``` python
class Xxx():
    def __int__(self):
        print("Int called")
        return 2
    
    def __float__(self):
        print("float called")
        return 2.2

a = Xxx()
df = pd.DataFrame({'aaa': [0, 1, 2, 3], 
              'bbb': [0.0, 1.1, 2.2, 3.3]
             })
df.loc[df['aaa'] == a]
# No result

df.loc[df['aaa'] == int(a)]
# Works

df.loc[df['bbb'] == a]
# No result

df.loc[df['bbb'] == float(a)]
# Works
```

How this could work (i'm not sure if it's a good idea / if there can be negative side-effects) is by inheriting from int / float.
Changing the above class to the following works fine for int tests above - and the corresponding alternative method works for float.

```python
class Xxx(int):
    def __int__(self):
        print("Int called")
        return 2
    
    def __float__(self):
        print("float called")
        return 2.2
```